pkg/protocol/block_dispatcher.go

```diff
@@ -78,18 +78,25 @@ func (b *BlockDispatcher) Dispatch(block *model.Block, src peer.ID) error {
 
 	matchingEngineFound := false
 	for _, engine := range []*engine.Engine{b.protocol.MainEngineInstance(), b.protocol.CandidateEngineInstance()} {
-		if engine != nil && !engine.WasShutdown() && (engine.ChainID() == slotCommitment.Chain().ForkingPoint.ID() || engine.BlockRequester.HasTicker(block.ID())) {
-			if b.inSyncWindow(engine, block) {
-				engine.ProcessBlockFromPeer(block, src)
-			} else {
-				// Stick too new blocks into the unsolid commitment buffer so that they can be dispatched once the
-				// engine instance is in sync (mostly needed for tests).
-				if !b.unsolidCommitmentBlocks.Add(slotCommitment.ID(), types.NewTuple(block, src)) {
-					return ierrors.Errorf("failed to add block %s to unsolid commitment buffer", block.ID())
+		e := engine
```

Comment:
Why not just assign in in the loop above? You assign it and keep using `engine` which shadows the package name

Reply:
The pointer will be reused in subsequent loop iterations, and we want to bind the correct value of the pointer in the defer, not the last one.

Reply:
```suggestion
		// Bind value of the pointer to a local variable, because the pointer in the loop will be reused and modified in the next iteration. That could be problematic with use of `defer`.
		e := engine
```